File: src/main/java/br/com/jiratorio/service/impl/JQLServiceImpl.java
package br.com.jiratorio.service.impl;

import br.com.jiratorio.domain.FluxColumn;
import br.com.jiratorio.domain.JQLBuilder;
import br.com.jiratorio.domain.entity.Board;
import br.com.jiratorio.service.JQLService;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

@Slf4j
@Service
public class JQLServiceImpl implements JQLService {

    @Override
    public String finalizedIssues(final Board board, final LocalDate start, final LocalDate end) {
        log.info("Method=finalizedIssues, board={}, startDate={}, endDate={}", board, start, end);

        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
        String startDate = start.format(formatter);
        String endDate = end.format(formatter) + " 23:59";
        FluxColumn fluxColumn = new FluxColumn(board);

        return new JQLBuilder()
                .append(" project = {project} ", board.getExternalId())
                .append(" AND ( STATUS CHANGED TO {endColumn} DURING({startDate}, {endDate}) ", board.getEndColumn(), startDate, endDate)
                .append("       OR ( STATUS CHANGED TO {lastColumn} DURING ({startDate}, {endDate}) AND NOT STATUS CHANGED TO {endColumn} )",
                        fluxColumn.getLastColumn(), startDate, endDate, board.getEndColumn())
                .append("       OR ( resolutiondate >= {startDate} AND resolutiondate <= {endDate} ", startDate, endDate)
                .append("              AND NOT STATUS CHANGED TO {lastColumn} AND NOT STATUS CHANGED TO {endColumn})",
                        fluxColumn.getLastColumn(), board.getEndColumn())
                .append("     ) ")
                .append(" AND issueType NOT IN ({issueTypes}) ", board.getIgnoreIssueType())
                .append(" AND status WAS IN ({startColumns}) ", fluxColumn.getStartColumns())
                .append(" AND status IN ({endColumns}) ", fluxColumn.getEndColumns())
                .build();
    }

    @Override
    public String openedIssues(final Board board) {
        log.info("Method=openedIssues, board={}", board);

        FluxColumn fluxColumn = new FluxColumn(board);

        return new JQLBuilder()
                .append(" project = {project} ", board.getExternalId())
                .append(" AND issueType NOT IN ({issueTypes}) ", board.getIgnoreIssueType())
                .append(" AND status IN ({fluxColumns}) ", fluxColumn.getWipColumns())
                .build();
    }

}

File: src/main/java/br/com/jiratorio/service/impl/JQLServiceImpl.kt
package br.com.jiratorio.service.impl

import br.com.jiratorio.domain.FluxColumn
import br.com.jiratorio.domain.entity.Board
import br.com.jiratorio.service.JQLService
import br.com.jiratorio.util.logger
import org.springframework.stereotype.Service
import java.time.LocalDate
import java.time.format.DateTimeFormatter

@Service
class JQLServiceImpl : JQLService {

    val log = logger()

    override fun finalizedIssues(board: Board, start: LocalDate, end: LocalDate): String {
        log.info("Method=finalizedIssues, board={}, startDate={}, endDate={}", board, start, end)

        val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")
        val startDate = start.format(formatter)
        val endDate = end.format(formatter) + " 23:59"
        val endColumn = board.endColumn
        val project = board.externalId

        val fluxColumn = FluxColumn(board)
        val lastColumn = fluxColumn.lastColumn
        val startColumns = fluxColumn.startColumns.joinToString(",") { "'$it'" }
        val endColumns = fluxColumn.endColumns.joinToString(",") { "'$it'" }

        val ignoredIssueTypes = board.ignoreIssueType?.joinToString(",", "AND issueType NOT IN (", ")") {
            "'$it'"
        } ?: ""

        return """
              | project = '$project'
              | AND (
              |      STATUS CHANGED TO '$endColumn' DURING('$startDate', '$endDate')
              |      OR (
              |          STATUS CHANGED TO '$lastColumn' DURING ('$startDate', '$endDate')
              |          AND NOT STATUS CHANGED TO '$endColumn'
              |      )
              |      OR (
              |          resolutiondate >= '$startDate' AND resolutiondate <= '$endDate'
              |          AND NOT STATUS CHANGED TO '$lastColumn'
              |          AND NOT STATUS CHANGED TO '$endColumn'
              |      )
              | )
              | $ignoredIssueTypes
              | AND status WAS IN ($startColumns)
              | AND status IN ($endColumns)
        """.trimMargin()
    }

    override fun openedIssues(board: Board): String {
        log.info("Method=openedIssues, board={}", board)

        val wipColumns = FluxColumn(board).wipColumns.joinToString(",") { "'$it'" }
        val project = board.externalId

        val ignoredIssueTypes = board.ignoreIssueType?.joinToString(",", "AND issueType NOT IN (", ")") {
            "'$it'"
        } ?: ""

        return """
            project = '$project'
            AND status IN ($wipColumns)
            $ignoredIssueTypes
        """.trimMargin()
    }

}
